modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsManagerImpl.java

```diff
@@ -87,195 +146,420 @@ public IgniteStatisticsRepository statisticsRepository() {
      *
-     * @param tbl Table to aggregate statistics by.
-     * @param selectedCols Columns to aggregate statistics by.
-     * @param tblPartStats Collection of partition level statistics.
-     * @return Local level statistics.
+     * @param data Global statistics to cache.
      */
-    private ObjectStatisticsImpl aggregateLocalStatistics(
-            GridH2Table tbl,
-            Column[] selectedCols,
-            Collection<ObjectPartitionStatisticsImpl> tblPartStats
-    ) {
-        Map<Column, List<ColumnStatistics>> colPartStats = new HashMap<>(selectedCols.length);
-        long rowCnt = 0;
-        for (Column col : selectedCols)
-            colPartStats.put(col, new ArrayList<>());
-
-        for (ObjectPartitionStatisticsImpl partStat : tblPartStats) {
-            for (Column col : selectedCols) {
-                ColumnStatistics colPartStat = partStat.columnStatistics(col.getName());
-                if (colPartStat != null) {
-                    colPartStats.compute(col, (k, v) -> {
-                        v.add(colPartStat);
-                        return v;
-                    });
-                }
+    public void saveGlobalStatistics(Collection<StatisticsObjectData> data) {
+        data.forEach(objData -> {
+            try {
+                ObjectStatisticsImpl objStat = StatisticsUtils.toObjectStatistics(this.ctx, objData);
+
+                statsRepos.saveGlobalStatistics(new StatisticsKey(objData.key().schema(), objData.key().obj()), objStat);
+            }
+            catch (IgniteCheckedException e) {
+                if (log.isDebugEnabled())
+                    log.debug(String.format("Cannot read global statistics %s", objData.key()));
             }
-            rowCnt += partStat.rowCount();
+        });
+    }
+
+    /**
+     * Register
+     *
+     * @param gatId Gathering id.
+     * @param data Collected statistics.
+     */
+    public void registerLocalResult(UUID gatId, Map<StatisticsObjectData, int[]> data) {
+        StatisticsGatheringContext stCtx = currColls.get(gatId);
+        if (stCtx == null) {
+            if (log.isDebugEnabled())
+                log.debug(String.format("Unable to register outdated statistics collection result %s", gatId));
+
+            return;
         }
+        Map<StatisticsKeyMessage, Collection<ObjectStatisticsImpl>> keyStats = new HashMap<>();
+        for (StatisticsObjectData objData : data.keySet()) {
+            keyStats.compute(objData.key(), (k, v) -> {
+                if (v == null)
+                    v = new ArrayList<>();
 
-        Map<String, ColumnStatistics> colStats = new HashMap<>(selectedCols.length);
-        for (Column col : selectedCols) {
-            ColumnStatistics stat = ColumnStatisticsCollector.aggregate(tbl::compareValues, colPartStats.get(col));
-            colStats.put(col.getName(), stat);
+                try {
+                    ObjectStatisticsImpl objStat = StatisticsUtils.toObjectStatistics(ctx, objData);
```

Comment:
is there any reason to convert this insert a compute's closure?

Reply:
outdated